kong/llm/init.lua

```diff
@@ -252,6 +381,159 @@ function _M.is_compatible(request, route_type)
+    end
+  end
+
+  -- send it to the ai service
+  local res, _, err, httpc = self.driver.subrequest(request, self.conf, self.http_opts, true)
+  if err then
+    return internal_server_error("failed to connect to " .. self.conf.model.provider .. " for streaming: " .. err)
+  end
+  if res.status ~= 200 then
+    err = "bad status code whilst opening streaming to " .. self.conf.model.provider .. ": " .. res.status
+    ngx.log(ngx.WARN, err)
+    return bad_request(err)
+  end
+
+  -- get a big enough buffered ready to make sure we rip the entire chunk(s) each time
+  local reader = res.body_reader
+  local buffer_size = 35536
+  local events
+
+  -- we create a fake "kong response" table to pass to the telemetry handler later
+  local telemetry_fake_table = {
+    response = buf:new(),
+    usage = {
+      prompt_tokens = prompt_tokens,
+      completion_tokens = 0,
+      total_tokens = 0,
+    },
+  }
+
+  ngx.status = 200
+  ngx.header["Content-Type"] = "text/event-stream"
+  ngx.header["Via"] = meta._SERVER_TOKENS
+
+  for k, v in pairs(res.headers) do
+    if not streaming_skip_headers[lower(k)] then
+      ngx.header[k] = v
+    end
+  end
+
+  -- server-sent events should ALWAYS be chunk encoded.
+  -- if they aren't then... we just won't support them.
+  repeat
+    -- receive next chunk
+    local buffer, err = reader(buffer_size)
+    if err then
+        ngx.log(ngx.ERR, err)
+        break
+    end
+
+    if buffer then
+      -- we need to rip each message from this chunk
+      events = {}
+      for s in buffer:gmatch("[^\r\n]+") do
+        table.insert(events, s)
+      end
+
+      local metadata
+
+      -- then parse each into the standard inference format
+      for i, event in ipairs(events) do
+        local event_t
+        local token_t
+
+        -- some LLMs do a final reply with token counts, and such
+        -- so we will stash them if supported
+        local formatted, err, this_metadata = self.driver.from_format(event, self.conf.model, "stream/" .. self.conf.route_type)
+        if err then
+          return internal_server_error(err)
+        end
+
+        metadata = this_metadata or metadata
+
+        -- handle event telemetry
+        if self.conf.logging.log_statistics then
+
+          if not ai_shared.streaming_has_token_counts[self.conf.model.provider] then
+            event_t = cjson.decode(formatted)
+            token_t = get_token_text(event_t)
+
+            -- incredibly loose estimate based on https://help.openai.com/en/articles/4936856-what-are-tokens-and-how-to-count-them
+            -- but this is all we can do until OpenAI fixes this...
+            --
+            -- essentially, every 4 characters is a token, with minimum of 1 per event
+            telemetry_fake_table.usage.completion_tokens =
+                telemetry_fake_table.usage.completion_tokens + math.ceil(#string.gsub(token_t, "^%s*(.-)%s*$", "%1") / 4)
```

Comment:
`kong.tools.utils.string.strip` has this covered:

https://github.com/Kong/kong/blob/c45adf9a598065f7fb84820e04f9b3498dae4893/kong/tools/string.lua#L27-L37

Reply:
Fixed in 6e1dc182f60460b0ecf4f37459fed2ca0b73f919